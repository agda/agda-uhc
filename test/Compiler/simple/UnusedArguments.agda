{-# OPTIONS -v treeless.opt:20 -v treeless.opt.unused:30 #-}

module _ where

open import Common.Prelude

-- First four arguments are unused.
maybe : ∀ {a b} {A : Set a} {B : Set b} → B → (A → B) → Maybe A → B
maybe z f nothing = z
maybe z f (just x) = f x

mapMaybe : ∀ {a b} {A : Set a} {B : Set b} → (A → B) → Maybe A → Maybe B
mapMaybe f x = maybe nothing (λ y → just (f y)) x

maybeToNat : Maybe Nat → Nat
maybeToNat m = maybe 0 (λ x → x) m

foldr : {A B : Set} → (A → B → B) → B → List A → B
foldr f z [] = z
foldr f z (x ∷ xs) = f x (foldr f z xs)

main : IO Unit
main = printNat (maybeToNat (just 42))
    ,, printNat (maybeToNat (mapMaybe (10 +_) (just 42)))
    ,, printNat (foldr _+_ 0 (1 ∷ 2 ∷ 3 ∷ 4 ∷ []))
