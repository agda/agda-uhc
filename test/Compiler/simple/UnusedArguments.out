EXECUTED_PROGRAM

ret > ExitSuccess
out > -- used args: _ _ _ _ e f g
out > UnusedArguments.maybe[stripped] =
out >   λ a b c →
out >     case c of
out >       Common.Maybe.Maybe.nothing → a
out >       Common.Maybe.Maybe.just d → b d
out > UnusedArguments.maybe =
out >   λ a b c d e f g →
out >     case g of
out >       Common.Maybe.Maybe.nothing → e
out >       Common.Maybe.Maybe.just h → f h
out > -- used args: _ _ _ _ e f
out > UnusedArguments.mapMaybe[stripped] =
out >   λ a b →
out >     UnusedArguments.maybe
out >       _ _ _ _ Common.Maybe.Maybe.nothing
out >       (λ c → Common.Maybe.Maybe.just (a c)) b
out > UnusedArguments.mapMaybe =
out >   λ a b c d e f →
out >     UnusedArguments.maybe
out >       _ _ _ _ Common.Maybe.Maybe.nothing
out >       (λ g → Common.Maybe.Maybe.just (e g)) f
out > UnusedArguments.maybeToNat =
out >   λ a → UnusedArguments.maybe _ _ _ _ 0 (λ b → b) a
out > -- used args: _ _ c d e
out > UnusedArguments.foldr[stripped] =
out >   λ a b c →
out >     case c of
out >       Agda.Builtin.List.List.[] → b
out >       Agda.Builtin.List.List._∷_ d e →
out >         a d (UnusedArguments.foldr _ _ a b e)
out > UnusedArguments.foldr =
out >   λ a b c d e →
out >     case e of
out >       Agda.Builtin.List.List.[] → d
out >       Agda.Builtin.List.List._∷_ f g →
out >         c f (UnusedArguments.foldr _ _ c d g)
out > UnusedArguments.main =
out >   Common.IO.then
out >     () () _ _
out >     (Common.IO.printNat
out >        (UnusedArguments.maybeToNat (Common.Maybe.Maybe.just 42)))
out >     (Common.IO.then
out >        () () _ _
out >        (Common.IO.printNat
out >           (UnusedArguments.maybeToNat
out >              (UnusedArguments.mapMaybe
out >                 _ _ _ _ (_+_ 10) (Common.Maybe.Maybe.just 42))))
out >        (Common.IO.printNat
out >           (UnusedArguments.foldr
out >              _ _ _+_ 0
out >              (Agda.Builtin.List.List._∷_
out >                 1
out >                 (Agda.Builtin.List.List._∷_
out >                    2
out >                    (Agda.Builtin.List.List._∷_
out >                       3 (Agda.Builtin.List.List._∷_ 4 Agda.Builtin.List.List.[])))))))
out > 425210
